walkeeper/src/timeline.rs

```diff
@@ -397,24 +432,70 @@ impl GlobalTimelines {
 
 #[derive(Debug)]
 struct FileStorage {
-    control_file: File,
+    // file used to prevent concurrent safekeepers running on the same data
+    lock_file: File,
+    // save timeline dir to avoid reconstructing it every time
+    timeline_dir: PathBuf,
     conf: SafeKeeperConf,
     persist_sync_control_file_seconds: Histogram,
     persist_nosync_control_file_seconds: Histogram,
 }
 
 impl Storage for FileStorage {
+    // persists state durably to underlying storage
+    // for description see https://lwn.net/Articles/457667/
     fn persist(&mut self, s: &SafeKeeperState, sync: bool) -> Result<()> {
         let _timer = if sync {
             &self.persist_sync_control_file_seconds
         } else {
             &self.persist_nosync_control_file_seconds
         }
         .start_timer();
-        self.control_file.seek(SeekFrom::Start(0))?;
-        s.ser_into(&mut self.control_file)?;
+
+        // write data to safekeeper.control.partial
+        let control_partial_path = self.timeline_dir.join(CONTROL_FILE_NAME_PARTIAL);
+        let mut control_partial = File::create(&control_partial_path).with_context(|| {
+            format!(
+                "failed to create partial control file at: {}",
+                &control_partial_path.display()
+            )
+        })?;
+        let mut buf = s.ser().with_context(|| "failed to serialize state")?;
+        // calculate checksum before resize
+        let checksum = crc32c::crc32c(&buf);
+
+        buf.resize(buf.len() + CHECKSUM_SIZE, 0u8);
+
+        let len = buf.len(); // overcome borrowing error
+        buf[len - CHECKSUM_SIZE..].copy_from_slice(&checksum.to_le_bytes());
+
+        control_partial.write_all(&buf).with_context(|| {
+            format!(
+                "failed to write safekeeper state into control file at: {}",
+                control_partial_path.display()
+            )
+        })?;
```

Comment:
I'd just call `write_all` twice or use `extend_from_slice`.

Reply:
write_all returns a result, but.resize and copy_from_slice are infallible (sort of), so I decided to have one place to handle the error. Do not see much of a difference

Reply:
Good point. `extend_from_slice` (if it does what I think it does) still should be easier to read than three lines of `resize`/`copy_from_slice`